### PR TITLE
added back end / front end checks and tests for token field length (closes #168)

### LINKED ISSUE #168: (Corpus Creation) Check that tokens are not larger than what can be stored in the DB
**Describe the bug**
Currently, tokens can be no longer than 64 chars. This means input cannot validate `\S{65, }`. If an error happens

**To Reproduce**
1. Go to New Corpus
2. Fill form
3. Add a token like Pseudosevangeliquolipapistoranabaptistiogiesuitanorbiterondepuritain
4. See error with no details

**Expected behavior**
1. Before submission, a message should be shown about the textarea having a wrong token. Maybe scroll the cursor directly to the problematic line.
2. If it is submitted, because Javascript failed or whatever, the backend should check that tokens are not longer than expected or at least raise a better error.

### PR DESCRIPTION
**Pull request by carinedengler**
Hi,

I'm opening the pull request as a WIP for the back end part first, I wanted to confirm that this is what you had in mind, and if so, for which fields the validation should be done. Also, I was wondering if the same validation shouldn't be done for any other field having a length restriction in the database schema which is set using user input, such as the corpus name.

### CODE REVIEW

File: app/utils/__init__.py [hunk 1]
```diff
@@ -1,8 +1,32 @@
...
 from .forms import string_to_none
+from typing import Dict
 
 
 class PyrrhaError(Exception):
     """Raised when errors specific to this application occur."""
     pass
+
+
+class ValidationError(Exception):
```
Comment (PonteIneptique) [inline app/utils/__init__.py:12]: Would it make sense to derive it from PyrrhaError ?

File: app/models/corpus.py [hunk 2]
```diff
@@ -870,6 +871,9 @@ def add_batch(corpus_id, word_tokens_dict, context_left=None, context_right=None
                 corpus=corpus_id,
                 order_id=i+1  # Asked by JB Camps...
             )
+            list(
+                validate_length(k, v, lengths={"form": 64}) for k, v in wt.items()
```
Comment (PonteIneptique) [inline app/models/corpus.py:875]: I like the current system you built. The only thing I'd like more is to show the line number potentially ? But that would mean getting the error and showing a modified message

**Comment by PonteIneptique** [comment]
Looks good to me :)  As for other fields, I see you have set up what's needed. I think there is less chances of a size constraint on other fields, and I wonder how the iteration over every key is gonna hurt performances. I'd probably feed the dictionary instead of key, value and iterate over the lengths Dict to avoid iterating on every object properties.

**Comment by carinedengler** [comment]
hi, I updated the pull request with the front end checks and integrated your suggestions into the back end checks

**Comment by PonteIneptique** [comment]
LGTM

**Comment by MrGecko** [comment]
Looks good to me though I didn't see the error message because the top of the page where it is inserted is off screen. 
I then clicked multiple times to understand and (logically) the same message kept piling up.

**Comment by carinedengler** [comment]
I was wondering just that about the error message - I put it where all of the other error messages are to be consistent with the rest of the interface, but made the focus on the line causing the error, I was wondering if an pop-up window would be more appropriate?

**Comment by PonteIneptique** [comment]
Maybe a scroll to the message error would be simpler / cleaner ?

**Comment by carinedengler** [comment]
after some quick tests, it seems that setSelectionRange only works when the focus is on the textArea, which is lost when the window is scrolled up

intuitively, I think it's more user-friendly when the problematic line is highlighted for the user, but I can also understand if you prefer a consistent user interface
(either way I will look into erasing the error message so that it will not pile up)

**Comment by PonteIneptique** [comment]
I think scrolling to the problematic line is way better. If you find a solution for this, I'd be okay with it.

**Comment by MrGecko** [comment]
Scrolling to the problematic line is ok (does it work with multiple errors too?) yet the user need to understand that is an issue. I would suggest to put the error message much closer to the input form section.

**Comment by carinedengler** [comment]
> Scrolling to the problematic line is ok (how does it behave with multiple errors?)
the error is raised as soon as the first line violating the constraint is found, so the next error will be shown when the form is re-submitted

**Comment by carinedengler** [comment]
> Scrolling to the problematic line is ok (how does it behave with multiple errors?)

it highlights the last line containing an error and shows error messages for each faulty line (order needs to be fixed), I'm not sure if that's the best behavior though, I can think of two possible solutions

* (my deleted comment referred to this) only highlight one erroneous line at a time when submitting the form to mirror back end behavior
* highlight the first erroneous line, show error messages for each faulty line

**Comment by PonteIneptique** [comment]
I like the second option :)

**Comment by carinedengler** [comment]
I implemented the second solution, and also rebased on the current dev branch

**Comment by MrGecko** [comment]
Given the proposed implementation it would be very easy to add the faulty line number.

Thank your for the improvements, I'll merge this PR as soon as we get this (extra useful) info in the error message.

**Comment by carinedengler** [comment]
I added the line number

**Comment by PonteIneptique** [comment]
I think there might be a bug in the tests: https://travis-ci.org/github/hipster-philology/pyrrha/builds/732968766#L2778

**Comment by carinedengler** [comment]
I updated the error message in the test

**Review by MrGecko — APPROVED** [review]

**Pull request merged**